sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/noop/NoopDataSource.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.noop
+
+import java.util.Optional
+
+import org.apache.spark.sql.SaveMode
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.sources.DataSourceRegister
+import org.apache.spark.sql.sources.v2._
+import org.apache.spark.sql.sources.v2.writer._
+import org.apache.spark.sql.types.StructType
+
+/**
+ * This is no-op datasource. It does not do anything besides consuming its input.
+ * This can be useful for benchmarking or to cache data without any additional overhead.
+ */
+private[noop] class NoopDataSource
```

Comment:
How about putting this is in the test source tree if it's only for tests and benchmarks?

Reply:
It can be used not only for benchmarks but for dataframe caching and filling in [IO cache](https://docs.databricks.com/user-guide/databricks-io-cache.html).

Reply:
I see, but it's not used anywhere in Spark though. Is it needed in Spark?

Reply:
You can also run benchmarks on production code. I would prefer to keep in it in `main`.

Reply:
Can we remove `private[noop]` in this class?

Reply:
I removed `private[noop]` for `NoopDataSource`